File: hutton/qsort.hs
module Main where

qsort :: Ord a => [a] -> [a]
qsort [] = []
qsort (x:xs) = qsort smaller ++ [x] ++ qsort larger
  where
    smaller = [a | a <- xs, a <= x]
    larger  = [b | b <- xs, b > x]

main = do
  let result = qsort [0,9,2,6,7,2,7,2,1,2]
  print result

File: hutton/qsort.js
function qsort ([h, ...tail]) {
  return (h === undefined)
    ? []
    : qsort(lesser(h))
        .concat([h])
        .concat(qsort(larger(h)))
  function lesser (x) { return tail.filter(y => y <= x) }
  function larger (x) { return tail.filter(y => y > x) }
}

console.log(
  qsort([0, 9, 2, 6, 7, 2, 7, 2, 1, 2])
)
